File: frontend/js/pieces/piece.js
//  * ===== מודול כלי השחמט =====
//  * מגדיר את מבנה כלי השחמט
//  * כולל סוג, צבע, מיקום ודרגה
//  * כולל פונקציות בסיסיות לכלי
export class Piece {
  constructor(color,type, row, col, grade = 0) {
    this.color = color; // 'w' או 'b'
    this.type = type; // סוג הכלי (P, R, N, B, Q, K)
    this.row = row; // מיקום השורה של הכלי בלוח
    this.col = col; // מיקום העמודה של הכלי בלוח
    this.grade = grade; // דרגה של הכלי, 0 אם לא מוגדר
  }

  getValidMoves() {
    return [];
  }
    //  פונקציה שמחזירה מיקום כתיאור טקסטואלי
  get positionText() {
    return `row: ${this.row}, col: ${this.col}`;
  }
}

File: java-engine/src/main/java/com/chess/domain/model/Piece.java
package com.chess.domain.model;

import java.util.List;

public abstract class Piece {

    private final Color color;         
    private final PieceType type;      
    private int row;                   
    private int col;     

    public enum Color {
        WHITE, BLACK;
    }

    public enum PieceType {
        PAWN, ROOK, KNIGHT, BISHOP, QUEEN, KING
    }

    public Piece(Color color, PieceType type, int row, int col) {
        this.color = color;
        this.type = type;
        this.row = row;
        this.col = col;
    }

    // Getters
    public Color getColor() { return color; }
    public PieceType getType() { return type; }
    public int getRow() { return row; }
    public int getCol() { return col; }

    // Method to move the piece
    public void moveTo(int newRow, int newCol) {
        this.row = newRow;
        this.col = newCol;
    }

    // Abstract method: subclasses must implement their valid moves
    public abstract List<Move> getValidMoves(Board board);
}